go/engine/team_create.go

```diff
@@ -0,0 +1,332 @@
+	sigGenericKey, err := e.G().Keyrings.GetSecretKeyWithPrompt(ctx.SecretKeyPromptArg(sigSKA, "to create a new team"))
+	if err != nil {
+		return
+	}
+	if err = sigGenericKey.CheckSecretKey(); err != nil {
+		return
+	}
+	sigKey, ok := sigGenericKey.(libkb.NaclSigningKeyPair)
+	if !ok {
+		err = fmt.Errorf("got an unexpected key type for device signing key: %T", sigGenericKey)
+		return
+	}
+
+	encSKA := libkb.SecretKeyArg{
+		Me:      me,
+		KeyType: libkb.DeviceEncryptionKeyType,
+	}
+	encGenericKey, err := e.G().Keyrings.GetSecretKeyWithPrompt(ctx.SecretKeyPromptArg(encSKA, "to create a new team"))
+	if err != nil {
+		return
+	}
+	if err = encGenericKey.CheckSecretKey(); err != nil {
+		return
+	}
+	encKey, ok = encGenericKey.(libkb.NaclDHKeyPair)
+	if !ok {
+		err = fmt.Errorf("got an unexpected key type for device encryption key: %T", encGenericKey)
+		return
+	}
+
+	return
+}
+
+func generatePerTeamKeys() (sharedSecret []byte, signingKey libkb.NaclSigningKeyPair, encryptionKey libkb.NaclDHKeyPair, err error) {
+	// This is the magical secret key, from which we derive a DH keypair and a
+	// signing keypair.
+	sharedSecret, err = libkb.RandBytes(32)
+	if err != nil {
+		return
+	}
+	encryptionKey, err = libkb.MakeNaclDHKeyPairFromSecretBytes(derivedSecret(sharedSecret, libkb.TeamDHDerivationString))
+	if err != nil {
+		return
+	}
+	signingKey, err = libkb.MakeNaclSigningKeyPairFromSecretBytes(derivedSecret(sharedSecret, libkb.TeamEdDSADerivationString))
+	if err != nil {
+		return
+	}
+	return
+}
+
+type PerTeamSharedSecretBoxes struct {
+	Generation    int               `json:"generation"`
+	EncryptingKid string            `json:"encrypting_kid"`
+	Nonce         string            `json:"nonce"`
+	Prev          string            `json:"prev"`
+	Boxes         map[string]string `json:"boxes"`
+}
+
+func boxTeamSharedSecret(secret []byte, senderKey libkb.NaclDHKeyPair, recipients map[string]keybase1.PerUserKey) (*PerTeamSharedSecretBoxes, error) {
+	noncePrefix, err := libkb.RandBytes(20)
+	if err != nil {
+		return nil, err
+	}
+	// The counter starts at 1, because 0 will be the prev secretbox, which is
+	// omitted for the team root link, because this is the first shared key.
+	var counter uint32 = 1
+	boxes := make(map[string]string)
+	for username, recipientPerUserKey := range recipients {
+		recipientPerUserGenericKeypair, err := libkb.ImportKeypairFromKID(recipientPerUserKey.EncKID)
+		if err != nil {
+			return nil, err
+		}
+		recipientPerUserNaclKeypair, ok := recipientPerUserGenericKeypair.(libkb.NaclDHKeyPair)
+		if !ok {
+			return nil, fmt.Errorf("got an unexpected key type for recipient KID in sharedTeamKeyBox: %T", recipientPerUserGenericKeypair)
+		}
+		var nonce [24]byte
+		counterBytes := [4]byte{}
+		binary.BigEndian.PutUint32(counterBytes[:], counter)
+		copy(nonce[:20], noncePrefix)
+		copy(nonce[20:24], counterBytes[:])
+		// TODO: pack [ 1, <per-user-seqno>, <nonce-bottom-4-bytes>, <box of current shared team secret> ]
```

Comment:
Woops, I implemented it in the following 3 lines and then forgot to delete the comment.